### DISCUSSION TITLE
[Python] Challenges working with relation.value_counts

### DISCUSSION THREAD

**Discussion by ned2** [duckdb/duckdb#8259] [General]
I've been using the relational API for doing some data analysis, and ran into some friction points using the value_counts method, that I thought I would start a discussion around to see if there's any opportunities to improve usability.

Let's work with this relation from a CSV: 
```python
duckdb.install_extension("https")
duckdb.load_extension("https")
csv = duckdb.read_csv("https://www.stats.govt.nz/assets/Uploads/Annual-enterprise-survey/Annual-enterprise-survey-2021-financial-year-provisional/Download-data/annual-enterprise-survey-2021-financial-year-provisional-csv.csv")
```

In order to count the distinct `Variable_category` values and show in sorted order, I found this was the simplest way to achieve that:

```python
csv.value_counts("Variable_category").order("2 DESC")
```

Observations: 
1. in the `order` method, it would be great for readability to be able to specify the count column by name rather than using the column index `2`. The only way I've been able to do this is with the expression  `'"count(Variable_category)" DESC'` which is a bit awkward, especially as it requires remembering to use double quotes for the column name. 
    ```python
    csv.value_counts("Variable_category").order('"count(Variable_category)" DESC')
    ```
2. While trying to debug this, I ran into an issue that if the column has been aliased as `name`, DuckDB wraps double quotes around the column identifier, which adds another issue:

    ```python
    new_rel = duckdb.sql("SELECT Variable_category AS name FROM csv")
    new_rel.value_counts("name").order('"count(name)" DESC')
    ```
    ```
    BinderException: Binder Error: Referenced column "count(name)" not found in FROM clause!
    Candidate bindings: "unnamed_relation_5acc7278d3c6812d.count("name")"
    ```
3. Would it be worth considering having some kind of a `sort` parameter like the Pandas `value_counts` method does, which would make the returned results sorted by counts by default to remove friction?

**Comment by Alex-Monahan** [comment]
This is an untested snippet, but maybe the `ORDER BY ALL` SQL concept could work here too?
https://duckdb.org/docs/sql/query_syntax/orderby#order-by-all

**Reply by ned2** [reply]
I did try that actually. In this case, I can't see a way that it would help, as we need to sort by the second column first, as it has the counts, then the names of the values would need to be sorted next. The `ALL` keyword sorts from left to right though, so it sorts by the value-names first.

Unless I'm missing a piece of the the puzzle!

**Reply by Tishj** [reply]
It's possible to add a `sort` parameter, but I think this is kind of a band-aid solution, imagine you want to also project other columns together with the aggregate.

I think the Expression API is a more scalable solution for this, I imagine we could support something like:
```py
csv.select(count("Variable_category").alias("name")).order(column("name").desc())
```

**Comment by Tishj** [comment]
```py
import duckdb
from duckdb import ColumnExpression
from duckdb import FunctionExpression

duckdb.execute("SET autoinstall_known_extensions=1")
duckdb.execute("SET autoload_known_extensions=1")

col = ColumnExpression("Variable_category")
count = FunctionExpression("count", col)

csv = duckdb.read_csv("https://www.stats.govt.nz/assets/Uploads/Annual-enterprise-survey/Annual-enterprise-survey-2021-financial-year-provisional/Download-data/annual-enterprise-survey-2021-financial-year-provisional-csv.csv")
rel = csv.project(
	count,
	col,
	groups="Variable_category"
)
rel.show()
```

`groups` should accept a list of Expressions, but that's not implemented yet
then it would be `groups=[col]`
